### ISSUE TITLE
support YAML for app declaration

### ISSUE THREAD

**Issue by rlidwka** [Unitech/pm2#578]
I'm looking at [this](https://github.com/meteor/meteor/issues/1116#issuecomment-48982351), and just realizing that is not a meteor issue. It's pm2 issue. JSON just doesn't work quite right when used for config files.

If we add yaml support, @dandv 's example could be rewritten like this:

```yaml
name: frontend
script: main.js

env:
  NODE_ENV: production
  METEOR_SETTINGS: |
    {
      "Kadira": {
         "appId": "...",
         "appSecret": "..."
      },
      "public": {
        "analytics_api_key": "xxx",
        ...
      }
    }
```

... solving all issues with escaping quotes.

I don't even mention that in production environments those files need to be commented out (which json doesn't support), or a handy feature of yaml that allows to define constants and reuse it anywhere in the same file (say, define "max_restarts" once and use that value for every app).

This would also remove the need for single quotes in "args" (and therefore eval(), see #568, which one could see as a possible security issue), replacing that with double quotes and a regular JSON.parse().

**Comment by naggie** [comment]
Yaml support would still be nice. May I add it?

**Comment by jshkurti** [comment]
Sure :)

**Comment by Unitech** [comment]
Yes would be nice, @naggie do you know any lightweight yaml library for node? (without too many dependencies)

**Comment by soyuka** [comment]
@Unitech I'm using https://www.npmjs.com/package/yamljs

**Comment by michaelBenin** [comment]
I would like to +1000 this. I like having comments in my configuration file.

**Comment by naggie** [comment]
Great, I'll do it this evening. @michaelBenin BTW I think you can put in comments anyway with the current JSON5 support: http://json5.org/. Nonetheless, YAML files are nicer for humans to read/write; also allowing anchors & references.

**Comment by naggie** [comment]
I have 2 options:

1. Drop JSON5 support in favour of YAML and risk the PR not being accepted. Bear in mind that JSON is a subset of YAML so we can still use JSON files; even commented JSON files
2. Change parser by file extension. This might not be straightforward as files are loaded into strings and passed around in CLI.js sometimes.


I would prefer option (1)

N.B. I noticed a few instances that might not work with the JSON5 parser -- a filename is given instead of the string

**Comment by soyuka** [comment]
You can also test the first character of the string that's not a space or a
new line. If it's `{` or `[` it's json if not it's yaml. Solution one, will
indeed bring a compatibility issue and therefore won't be merged...
If a file is sent as an argument testing the file extension seem fair to me.
Le mar. 31 mars 2015 à 10:20, Callan Bryant <notifications@github.com> a
écrit :

**Comment by felixfbecker** [comment]
I also think PM2 should accept multiple file types for app declarations. ***It currently supports embedding JS inside JSON!*** If you write Javascript inside a "JavaScript Object Notation" file you might as well just put in a JS file with `module.exports = `. PM2 should accept .yaml, .json and .js.
JSON really is bad without official support for comments and all the quoting going on.

**Comment by Unitech** [comment]
Feature landed in development branch, thanks to the lightweight YAML parsing module https://www.npmjs.com/package/yamljs

**Comment by Unitech** [comment]
Update available (PM2@1.1.1 - latest stable):

```
$ npm install pm2@latest -g
$ pm2 update
```

Please re-open this issue if you notice any problem,
